osu.Game.Tests/Visual/Gameplay/TestSceneCompletionCancellation.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using NUnit.Framework;
+using osu.Framework.Allocation;
+using osu.Framework.Audio;
+using osu.Framework.Audio.Track;
+using osu.Framework.Screens;
+using osu.Framework.Testing;
+using osu.Framework.Timing;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Osu;
+using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Storyboards;
+using osuTK;
+
+namespace osu.Game.Tests.Visual.Gameplay
+{
+    public class TestSceneCompletionCancellation : PlayerTestScene
+    {
+        private Track track;
+
+        [Resolved]
+        private AudioManager audio { get; set; }
+
+        protected override bool AllowFail => false;
+
+        public TestSceneCompletionCancellation()
+            : base(new OsuRuleset())
+        {
+        }
+
+        [SetUpSteps]
+        public override void SetUpSteps()
+        {
+            base.SetUpSteps();
+
+            // Ensure track has actually running before attempting to seek
+            AddUntilStep("wait for track to start running", () => track.IsRunning);
+        }
+
+        [Test]
+        public void TestCancelCompletionOnRewind()
+        {
+            cancelCompletionSteps();
+
+            AddAssert("no attempt to push ranking", () => !((FakeRankingPushPlayer)Player).GotoRankingInvoked);
+        }
+
+        [Test]
+        public void TestReCompleteAfterCancellation()
+        {
+            cancelCompletionSteps();
+
+            // Attempt completing again.
+            AddStep("seek to completion again", () => track.Seek(5000));
+            AddUntilStep("completion set by processor", () => Player.ScoreProcessor.HasCompleted.Value);
+
+            AddWaitStep("wait", 5);
+
+            AddAssert("attempted to push ranking", () => ((FakeRankingPushPlayer)Player).GotoRankingInvoked);
+        }
+
+        /// <summary>
+        /// Tests whether can still pause after cancelling completion
+        /// by reverting <see cref="IScreen.ValidForResume"/> back to true.
+        /// </summary>
+        [Test]
+        public void TestCanPauseAfterCancellation()
+        {
+            cancelCompletionSteps();
+
+            AddStep("pause", () => Player.Pause());
+            AddAssert("paused successfully", () => Player.GameplayClockContainer.IsPaused.Value);
+        }
+
+        private void cancelCompletionSteps()
+        {
+            AddStep("seek to completion", () => track.Seek(5000));
+            AddUntilStep("completion set by processor", () => Player.ScoreProcessor.HasCompleted.Value);
+
+            AddStep("rewind to cancel", () => track.Seek(4000));
+            AddUntilStep("completion cleared by processor", () => !Player.ScoreProcessor.HasCompleted.Value);
+
+            AddWaitStep("wait", 5);
```

Comment:
this seems prone to potential failure. can you make it an until step instead?

Reply:
I don't think I can, this is a step that waits a couple of seconds before asserting to ensure that there was no attempt to push ranking screen, I will describe that in the step for clarity.

Reply:
In that case please make a const in player (`RESULTS_DISPLAY_DELAY`) and base the wait step duration on `RESULTS_DISPLAY_DELAY / TimePerAction * 2` or similar (`* 2` is to add some degree of safety), and add a comment.

Reply:
Although with current values `(1000 / 200 * 2)`, `* 2` actually buffed up the wait count to `10`, I think ceiling is more appropriate here such that it doesn't fall one wait count behind.

Reply:
you may be able to reproduce an fail under visual tests due to the way scheduler works. make sure to test it thoroughly if you want to cut it that close.